desc/objectives/utils.py

```diff
@@ -81,19 +82,43 @@ def factorize_linear_constraints(constraints, objective):  # noqa: C901
         xz = _tree_zeros_like([t.params_dict for t in con.things])
         # computing A matrix for each constraint for each thing in the optimization
         for thing in objective.things:
-            if thing in con.things:
-                A_ = con.jac_scaled(*xz)[con.things.index(thing)]
+            if not isinstance(thing, OptimizableCollection):
+                if thing in con.things:
+                    A_ = con.jac_scaled(*xz)[con.things.index(thing)]
+                else:
+                    A_ = {
+                        arg: jnp.zeros((con.dim_f, dimx))
+                        for arg, dimx in thing.dimensions.items()
+                    }
+                args = (
+                    objective._args
+                    if prox_flag and isinstance(thing, Equilibrium)
+                    else thing.optimizable_params
+                )
+                A_per_thing.append(jnp.hstack([A_[arg] for arg in args]))
+
             else:
-                A_ = {
-                    arg: jnp.zeros((con.dim_f, dimx))
-                    for arg, dimx in thing.dimensions.items()
-                }
-            args = (
-                objective._args
-                if prox_flag and isinstance(thing, Equilibrium)
-                else thing.optimizable_params
-            )
-            A_per_thing.append(jnp.hstack([A_[arg] for arg in args]))
+                A_per_subthing = []
+                # it is an optimizable collection, loop through its sub-objects
+                for subthing in thing:
+                    if subthing in con._subthings:
+                        # this returns a tuple of length 1 for some reason?
+                        A_ = con.jac_scaled(*xz)[0][con._subthings.index(subthing)]
```

Comment:
not sure why I needed a 0 index here, but it `con.jac_scaled(*xz)` was returning a tuple of length 1 that contained a list of dicts of length(con._subthings), so I had to index into the tuple here. I am really tired so likely i just am missing something stupid here

Reply:
iirc `con.jac_scaled` returns a tuple of dict of array, where the tuple corresponds to how many sets of params the constraint takes, and each dict is the derivative wrt that set of params